File: templates/skills/dispatching-parallel-agents/scripts/result_collector.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
结果回收与冲突检测器。

并行子 Agent 完成后，回收结果、检测冲突、决定是否需要顺序降级。

用法:
    python3 result_collector.py collect --group-id batch1
    python3 result_collector.py check-conflicts --group-id batch1
"""

from __future__ import annotations

import argparse
import json
import os
import subprocess
import sys
from datetime import datetime
from typing import Any, Dict, List, Optional

sys.path.insert(0, os.path.dirname(__file__))
from agent_registry import read_registry, write_registry, get_group_status


# =============================================================================
# Conflict Detection
# =============================================================================


def _detect_branch_for_agent(agent: Dict[str, Any]) -> Optional[str]:
    """Resolve the agent branch from its worktree instead of assuming a name."""
    worktree_path = agent.get("worktree_path")
    if not worktree_path:
        return None

    result = subprocess.run(
        ["git", "rev-parse", "--abbrev-ref", "HEAD"],
        capture_output=True,
        text=True,
        encoding="utf-8",
        errors="replace",
        cwd=worktree_path,
    )
    if result.returncode != 0:
        return None

    branch = result.stdout.strip()
    if not branch or branch == "HEAD":
        return None
    return branch


def _check_merge_compatibility(root: str, branch: str) -> Dict[str, Any]:
    """Check merge compatibility using commit graph state only.

    `git merge-tree --write-tree` merges the named commits without touching the
    caller's working tree, so unrelated local modifications cannot trigger
    false conflicts.
    """
    merge_result = subprocess.run(
        ["git", "merge-tree", "--write-tree", "--quiet", "HEAD", branch],
        capture_output=True,
        text=True,
        encoding="utf-8",
        errors="replace",
        cwd=root,
    )
    return {
        "ok": merge_result.returncode == 0,
        "stderr": merge_result.stderr.strip(),
        "stdout": merge_result.stdout.strip(),
    }


def detect_merge_conflicts(
    group_id: str,
    project_root: Optional[str] = None,
) -> Dict[str, Any]:
    """Detect merge conflicts between worktree branches in a group.

    For each completed agent in the group, checks if its branch
    can be merged without conflicts.
    """
    root = project_root or os.getcwd()
    registry = read_registry(root)
    group = registry.get("groups", {}).get(group_id)

    if not group:
        return {"error": f"Group {group_id} not found"}

    agents = [
        registry["agents"].get(aid)
        for aid in group["agent_ids"]
        if aid in registry.get("agents", {})
    ]

    conflicts = []
    clean_merges = []

    for agent in agents:
        if not agent or agent["status"] != "completed":
            continue

        branch = _detect_branch_for_agent(agent)
        if not branch:
            clean_merges.append(agent["task_id"])
            continue

        result = _check_merge_compatibility(root, branch)

        if not result["ok"]:
            conflicts.append({
                "task_id": agent["task_id"],
                "agent_id": agent["agent_id"],
                "branch": branch,
                "error": (result.get("stderr") or result.get("error") or "")[:300],
            })
        else:
            clean_merges.append(agent["task_id"])

    has_conflicts = len(conflicts) > 0

    # Update group status
    if has_conflicts:
        group["conflict_detected"] = True
        write_registry(registry, root)

    return {
        "group_id": group_id,
        "has_conflicts": has_conflicts,
        "conflicts": conflicts,
        "clean_merges": clean_merges,
        "recommendation": "sequential_fallback" if has_conflicts else "merge_all",
    }


def run_aggregate_verification(
    commands: List[str],
    cwd: Optional[str] = None,
    timeout: int = 120,
) -> List[Dict[str, Any]]:
    """Run aggregate verification after parallel completion."""
    results = []
    for cmd in commands:
        try:
            result = subprocess.run(
                cmd, shell=True, capture_output=True,
                text=True, encoding="utf-8", errors="replace",
                timeout=timeout, cwd=cwd or os.getcwd(),
            )
            results.append({
                "command": cmd,
                "exit_code": result.returncode,
                "passed": result.returncode == 0,
                "output": (result.stdout or result.stderr or "")[:500],
            })
        except subprocess.TimeoutExpired:
            results.append({
                "command": cmd,
                "exit_code": -1,
                "passed": False,
                "output": f"Timeout after {timeout}s",
            })

    return results


# =============================================================================
# Result Collection
# =============================================================================


def collect_group_results(
    group_id: str,
    verification_commands: Optional[List[str]] = None,
    project_root: Optional[str] = None,
) -> Dict[str, Any]:
    """Collect results from all agents in a group.

    Implements SKILL.md Step 7 (Result Recovery) and Step 8 (Conflict Detection).
    """
    root = project_root or os.getcwd()
    group_status = get_group_status(group_id, root)

    if "error" in group_status:
        return group_status

    agents = group_status.get("agents", [])

    # Categorize results
    completed = [a for a in agents if a and a.get("status") == "completed"]
    failed = [a for a in agents if a and a.get("status") == "failed"]
    still_running = [a for a in agents if a and a.get("status") == "running"]

    result: Dict[str, Any] = {
        "group_id": group_id,
        "total_agents": len(agents),
        "completed": len(completed),
        "failed": len(failed),
        "still_running": len(still_running),
        "completed_tasks": [a["task_id"] for a in completed if a],
        "failed_tasks": [
            {"task_id": a["task_id"], "output": a.get("output_summary", "")}
            for a in failed if a
        ],
    }

    # If still running, return partial result
    if still_running:
        result["status"] = "partial"
        result["waiting_for"] = [a["task_id"] for a in still_running if a]
        return result

    # All done → check conflicts
    conflict_result = detect_merge_conflicts(group_id, root)
    result["conflicts"] = conflict_result

    # Run aggregate verification if provided
    if verification_commands:
        verification = run_aggregate_verification(verification_commands, root)
        result["verification"] = verification
        result["all_verified"] = all(v["passed"] for v in verification)
    else:
        result["all_verified"] = True

    # Determine overall status
    if conflict_result.get("has_conflicts"):
        result["status"] = "conflict"
        result["recommendation"] = "sequential_fallback"
    elif failed:
        result["status"] = "partial_failure"
        result["recommendation"] = "continue_with_failures"
    else:
        result["status"] = "success"
        result["recommendation"] = "merge_all"

    result["collected_at"] = datetime.now().isoformat()
    return result


# =============================================================================
# CLI
# =============================================================================


def main() -> int:
    parser = argparse.ArgumentParser(description="结果回收与冲突检测")
    sub = parser.add_subparsers(dest="command")

    # collect
    p_collect = sub.add_parser("collect", help="回收分组结果")
    p_collect.add_argument("--group-id", required=True)
    p_collect.add_argument("--verify", nargs="*", help="聚合验证命令")

    # check-conflicts
    p_conflict = sub.add_parser("check-conflicts", help="检测合并冲突")
    p_conflict.add_argument("--group-id", required=True)

    args = parser.parse_args()

    if args.command == "collect":
        result = collect_group_results(
            args.group_id,
            verification_commands=args.verify,
        )
    elif args.command == "check-conflicts":
        result = detect_merge_conflicts(args.group_id)
    else:
        parser.print_help()
        return 1

    print(json.dumps(result, indent=2, ensure_ascii=False))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: core/skills/dispatching-parallel-agents/scripts/result_collector.js
#!/usr/bin/env node

/** 并行分派结果收集器 —— 负责冲突检测、聚合验证和分组结果汇总 */

const { spawnSync } = require('child_process')
const { readRegistry, writeRegistry, getGroupStatus } = require('./agent_registry')

/**
 * 检测 agent 对应 worktree 的当前分支名
 * @param {object} agent - agent 条目（需含 worktree_path）
 * @returns {string|null} 分支名，无法检测时返回 null
 */
function detectBranchForAgent(agent) {
  const worktreePath = agent.worktree_path
  if (!worktreePath) return null

  const result = spawnSync('git', ['rev-parse', '--abbrev-ref', 'HEAD'], {
    cwd: worktreePath,
    encoding: 'utf8',
    maxBuffer: 10 * 1024 * 1024,
  })

  if (result.status !== 0) return null
  const branch = String(result.stdout || '').trim()
  if (!branch || branch === 'HEAD') return null
  return branch
}

/**
 * 使用 git merge-tree 检查指定分支与 HEAD 的合并兼容性
 * @param {string} root - 仓库根目录
 * @param {string} branch - 待检查的分支名
 * @returns {object} 包含 ok（是否无冲突）、stderr、stdout
 */
function checkMergeCompatibility(root, branch) {
  const result = spawnSync('git', ['merge-tree', '--write-tree', '--quiet', 'HEAD', branch], {
    cwd: root,
    encoding: 'utf8',
    maxBuffer: 10 * 1024 * 1024,
  })

  return {
    ok: result.status === 0,
    stderr: String(result.stderr || '').trim(),
    stdout: String(result.stdout || '').trim(),
  }
}

/**
 * 检测指定分组中已完成 agent 的合并冲突，有冲突时标记 group 并建议顺序降级
 * @param {string} groupId - 分组 ID
 * @param {string|null} projectRoot - 项目根目录
 * @returns {object} 冲突检测结果，含 has_conflicts、conflicts 列表和建议
 */
function detectMergeConflicts(groupId, projectRoot = null) {
  const root = projectRoot || process.cwd()
  const registry = readRegistry(root)
  const group = registry.groups?.[groupId]
  if (!group) return { error: `Group ${groupId} not found` }

  const agents = (group.agent_ids || []).map((agentId) => registry.agents?.[agentId]).filter(Boolean)
  const conflicts = []
  const cleanMerges = []

  for (const agent of agents) {
    if (agent.status !== 'completed') continue

    const branch = detectBranchForAgent(agent)
    if (!branch) {
      cleanMerges.push(agent.task_id)
      continue
    }

    const result = checkMergeCompatibility(root, branch)
    if (!result.ok) {
      conflicts.push({
        task_id: agent.task_id,
        agent_id: agent.agent_id,
        branch,
        error: (result.stderr || result.error || '').slice(0, 300),
      })
    } else {
      cleanMerges.push(agent.task_id)
    }
  }

  const hasConflicts = conflicts.length > 0
  if (hasConflicts) {
    group.conflict_detected = true
    writeRegistry(registry, root)
  }

  return {
    group_id: groupId,
    has_conflicts: hasConflicts,
    conflicts,
    clean_merges: cleanMerges,
    recommendation: hasConflicts ? 'sequential_fallback' : 'merge_all',
  }
}

/**
 * 在指定目录下依次执行验证命令，收集每条命令的通过状态和输出
 * @param {string[]} commands - 待执行的 shell 命令列表
 * @param {string|null} cwd - 工作目录
 * @param {number} timeout - 单条命令超时秒数
 * @returns {object[]} 每条命令的执行结果（含 passed、exit_code、output）
 */
function runAggregateVerification(commands, cwd = null, timeout = 120) {
  const root = cwd || process.cwd()
  const results = []

  for (const command of commands || []) {
    const result = spawnSync(command, [], {
      cwd: root,
      shell: true,
      encoding: 'utf8',
      timeout: timeout * 1000,
      maxBuffer: 10 * 1024 * 1024,
    })

    if (result.error && result.error.code === 'ETIMEDOUT') {
      results.push({
        command,
        exit_code: -1,
        passed: false,
        output: `Timeout after ${timeout}s`,
      })
      continue
    }

    const output = `${result.stdout || ''}${result.stderr || ''}`.slice(0, 500)
    results.push({
      command,
      exit_code: typeof result.status === 'number' ? result.status : -1,
      passed: result.status === 0,
      output,
    })
  }

  return results
}

/**
 * 汇总指定分组的执行结果：统计完成/失败/运行中数量，检测冲突，运行验证命令
 * @param {string} groupId - 分组 ID
 * @param {string[]|null} verificationCommands - 聚合验证命令列表
 * @param {string|null} projectRoot - 项目根目录
 * @returns {object} 汇总结果，含状态、冲突信息、验证结果和建议
 */
function collectGroupResults(groupId, verificationCommands = null, projectRoot = null) {
  const root = projectRoot || process.cwd()
  const groupStatus = getGroupStatus(groupId, root)
  if (groupStatus.error) return groupStatus

  const agents = groupStatus.agents || []
  const completed = agents.filter((agent) => agent?.status === 'completed')
  const failed = agents.filter((agent) => agent?.status === 'failed')
  const stillRunning = agents.filter((agent) => agent?.status === 'running')

  const result = {
    group_id: groupId,
    total_agents: agents.length,
    completed: completed.length,
    failed: failed.length,
    still_running: stillRunning.length,
    completed_tasks: completed.map((agent) => agent.task_id),
    failed_tasks: failed.map((agent) => ({ task_id: agent.task_id, output: agent.output_summary || '' })),
  }

  if (stillRunning.length > 0) {
    result.status = 'partial'
    result.waiting_for = stillRunning.map((agent) => agent.task_id)
    return result
  }

  const conflictResult = detectMergeConflicts(groupId, root)
  result.conflicts = conflictResult

  if (verificationCommands) {
    const verification = runAggregateVerification(verificationCommands, root)
    result.verification = verification
    result.all_verified = verification.every((item) => item.passed)
  } else {
    result.all_verified = true
  }

  if (conflictResult.has_conflicts) {
    result.status = 'conflict'
    result.recommendation = 'sequential_fallback'
  } else if (failed.length > 0) {
    result.status = 'partial_failure'
    result.recommendation = 'continue_with_failures'
  } else {
    result.status = 'success'
    result.recommendation = 'merge_all'
  }

  result.collected_at = new Date().toISOString()
  return result
}

function parseOption(args, flag) {
  const index = args.indexOf(flag)
  return index >= 0 ? args[index + 1] : undefined
}

function collectVerifyValues(args) {
  const index = args.indexOf('--verify')
  if (index < 0) return undefined
  const values = []
  for (let i = index + 1; i < args.length; i += 1) {
    if (args[i].startsWith('--')) break
    values.push(args[i])
  }
  return values
}

function hasFlag(args, flag) {
  return args.includes(flag)
}

function printJson(value) {
  process.stdout.write(`${JSON.stringify(value, null, 2)}\n`)
}

function main() {
  const args = process.argv.slice(2)
  const command = args.shift()

  if (command === 'collect') {
    if (!hasFlag(args, '--group-id')) {
      process.stderr.write('Missing required arguments\n')
      process.exitCode = 1
      return
    }
    printJson(collectGroupResults(parseOption(args, '--group-id'), collectVerifyValues(args)))
    return
  }

  if (command === 'check-conflicts') {
    if (!hasFlag(args, '--group-id')) {
      process.stderr.write('Missing required arguments\n')
      process.exitCode = 1
      return
    }
    printJson(detectMergeConflicts(parseOption(args, '--group-id')))
    return
  }

  process.stderr.write('Usage: node result_collector.js <collect|check-conflicts> ...\n')
  process.exitCode = 1
}

const _detect_branch_for_agent = detectBranchForAgent
const _check_merge_compatibility = checkMergeCompatibility
const detect_merge_conflicts = detectMergeConflicts
const run_aggregate_verification = runAggregateVerification
const collect_group_results = collectGroupResults

module.exports = {
  detectBranchForAgent,
  checkMergeCompatibility,
  detectMergeConflicts,
  runAggregateVerification,
  collectGroupResults,
  _detect_branch_for_agent,
  _check_merge_compatibility,
  detect_merge_conflicts,
  run_aggregate_verification,
  collect_group_results,
}

if (require.main === module) main()
